pallets/elections/src/lib.rs

```diff
@@ -392,26 +418,28 @@ pub mod pallet {
             Ok(())
         }
 
-        fn kick_out_underperformed_non_reserved_validators() {
-            let mut to_be_kicked = BTreeMap::from_iter(ToBeKickedOutFromCommittee::<T>::iter());
-            if !to_be_kicked.is_empty() {
-                let mut to_be_removed = Vec::new();
-                let mut filtered_non_reserved_in_order = Vec::new();
-                let non_reserved_validators = NextEraNonReservedValidators::<T>::get();
-                for non_reserved_validator in non_reserved_validators {
-                    match to_be_kicked.remove_entry(&non_reserved_validator) {
-                        Some(candidate_and_reason) => to_be_removed.push(candidate_and_reason),
-                        None => filtered_non_reserved_in_order.push(non_reserved_validator),
-                    }
-                }
-                if !to_be_removed.is_empty() {
-                    info!(target: "pallet_elections", "Removing following validators from non reserved, {:?}", to_be_removed);
-                    NextEraNonReservedValidators::<T>::put(filtered_non_reserved_in_order);
-                    Self::deposit_event(Event::KickOutValidators(to_be_removed));
-                }
-                let _result = ToBeKickedOutFromCommittee::<T>::clear(u32::MAX, None);
+        fn emit_fresh_bans_event() {
+            let active_era = <T as Config>::EraInfoProvider::active_era().unwrap_or(0);
+            let fresh_bans: Vec<_> = Banned::<T>::iter()
+                .filter(|(_acc, info)| info.start == active_era)
+                .collect();
+            if !fresh_bans.is_empty() {
+                info!(target: "pallet_elections", "Fresh bans in era {}: {:?}",active_era, fresh_bans);
+                Self::deposit_event(Event::BanValidators(fresh_bans));
             }
         }
+
+        fn unban() {
+            let active_era = <T as Config>::EraInfoProvider::active_era().unwrap_or(0);
+            let ban_period = BanConfig::<T>::get().ban_period;
+            let unban = Banned::<T>::iter().filter_map(|(account_id, ban_info)| {
+                if ban_info.start.saturating_add(ban_period) >= active_era {
```

Comment:
this also could suffer from that: `ban_info.start.saturating_add(ban_period) >= active_era.saturating_add(1)`

Reply:
The whole logic was moved to beginning of an era.